### DISCUSSION TITLE
I2Cdevice - PCF8591 and PCF8574

### DISCUSSION THREAD

**Discussion by alpha71** [pybricks/support#317] [Q&A]
Hello everybody,
i'm open this topic coz i wanna read the analog value from a i2c 8bit A/D D/A converter, PCF8591.
Now here is the point: the pcf8591 has 4 analog IN and one analog OUT, actually is commonly used in Arduino, but in the past MINDSENSOR sold a general purpose board for MINDSTORM NXT (a sort of kit!) using this chip (and also the digital version PCF8574, with 8 digital I/O). Actually i already used it with NXT and NXC programming languages, without much issues, now my targert is to use it with EV3 and micropython programming. Of course i'm not a proper programmer and surely i'm a "new user" in Python but i i'd like to experiment some generic sensor to propose in my robot class. Ok then:
i tried a software:

```python
from pybricks.hubs import EV3Brick
from pybricks.parameters import Port, Stop, Direction, Button, Color
from pybricks.tools import wait, StopWatch, DataLog
from pybricks.robotics import DriveBase
from pybricks.media.ev3dev import SoundFile, ImageFile
from pybricks.iodevices import I2CDevice

ev3 = EV3Brick()
# Initialize I2C Sensor
device = I2CDevice(Port.S2, 0x90 >> 1)
result, = device.read(reg=None, length=1)
while(True):
    device.write(reg=None, data=b'\x00') #write to the channel 00
    device.read(reg=None, length=1) #read 1bit from the channel 00
```

then after this "effort" i'm stucked! I don't know how to print the value (wich should assign to a variable!), whatever in the Console or thru the EV3 display, infact if i assign a variable such as analog0 = device.read(.....), it gives a error (of course). I guess i've to create a Class...but actually i'm not clear how to move on!
Maybe somebody already work around this issue in case please let me undestand how to do :)
Appreciate 

![pcf8591_addressing](https://user-images.githubusercontent.com/37776195/116383737-6bba0780-a817-11eb-97b6-5eb7b366500f.JPG)

**Comment by dlech** [comment]
What is the error that you see?

**Reply by alpha71** [reply]
Hello, 
meanwhile i reviewed the code and actually i found it works! :) at least i got a sort of output i just cannot understand how to convert it in a "number", such as 0-255 to dispaly them into EV3 display! Infact i got a part the hexadecimal also several simbols....
Moreover i'm still concerning if I'm doing right.... :) 
thank you again

Alessandro

![Cattura](https://user-images.githubusercontent.com/37776195/116475915-c33d8f00-a87a-11eb-98a3-045b1c69d353.JPG)


![i2c_01](https://user-images.githubusercontent.com/37776195/116475643-6c37ba00-a87a-11eb-9b37-3d0ab2b68825.JPG)

**Reply by dlech** [reply]
You can use `analog0[0]` to convert it to a number.

The `read` function returns a `bytes` object, which is basically an array of bytes, so `[0]` will get the first byte in the array.

**Reply by alpha71** [reply]
Hi David, 
that's great, it works! 
Thank you
Another question:
the PCF8591 has 4 output then if i wanna read them together or anyway with a quick switching of them, such as a MUX, how should i approach?

**Reply by dlech** [reply]
I see in the screenshot that there is an auto-increment flag. Maybe that could be of some use?

**Reply by alpha71** [reply]
Yeah. if set data = 0x04 i can get all 4 channels' signals, now i don't know how can discriminate them and print and/or use indipendently!

In the image can see i try to assign a variable to each one but actually can't work, because i'm going to write in the same loop 4 times the registries (stupid me!) 
What if i use Class and Def? Can i make one array?

![immagine](https://user-images.githubusercontent.com/37776195/116619078-483eac00-a940-11eb-896e-305e10a9c9b9.png)

**Comment by alpha71** [comment]
Here the board!

![IMG_0363](https://user-images.githubusercontent.com/37776195/116621837-1596b280-a944-11eb-9127-eab99bd02307.JPG)

**Reply by dlech** [reply]
What happens if you set `length=8`? These chips usually clock from the I2C line, so the first one or two values read are usually stale values (the data sheet should say something about this). Then you can ignore the first 4 values read and the last 4 should be channels 0 to 3 all measured at about the same time.

**Reply by alpha71** [reply]
Hello, no way! if i assign Lenght=8 i get error!

**Comment by alpha71** [comment]
I found a way to read and write on the chip assigning the registries as you can see in the picture. I just cannot catch why it's Reg=0x05 =b0100 if the registry should be 0x04. This happens in all input infact they are all 1bit more....Why?
In the written the registry it's right infact 0x40!!
Another question is:
how to avoid the scrolling of the screen, what I meant is if I display the value of the analog input (or any other value from sensors) i wanna refresh it just in the same line without scrolling. I didn0t see this in the documentation.

![immagine](https://user-images.githubusercontent.com/37776195/116776570-4f150e00-aa69-11eb-8a78-e0ba33357c5d.png)

**Reply by dlech** [reply]
> This happens in all input infact they are all 1bit more....Why?

This is what I was trying to explain. The analog chip is measuring and sending data at the same time. The new measurement cannot be finished before starting to send the bits. So the data that is sent is the old data that was measured during the last read and the new measurement will be saved until the next read. So for each read you get the measurement that was made during the last read.

>  i wanna refresh it just in the same line without scrolling

You can use [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code) to move the cursor. Maybe something like this:

    print('\x1B[4F\x1B[G')  # go up 4 lines and to column 1

**Reply by alpha71** [reply]
Hello David, first thank you for your help, it was precious. These days i move on to test several devices I2C and others , mainly from Mindsensor and HiTechnic (some succesfully others not :( ) and for these i'm going to open another discussion today (also include NXT class).
Thank again for your support
Ale
